universal-login-commons/lib/core/utils/securityCodes.ts

```diff
@@ -32,6 +29,28 @@ export const isValidCode = (code: number[], publicKey: string) => {
   return deepEqual(expectedCode, code);
 };
 
+export const findValidAddressFromCodePart = (enteredCode: number[], incomingPublicKeys: string[]) => {
+  const possibleAddresses = findPossibleAddressesFromCodePart(enteredCode, incomingPublicKeys);
+  return possibleAddresses.length === 1 ? possibleAddresses[0] : undefined;
+};
+
+export const findPossibleAddressesFromCodePart = (enteredCode: number[], incomingPublicKeys: string[]) => {
+  const generatedCodes = incomingPublicKeys.map((key) => generateCode(key));
+  const validIndicies = findIndiciesOfValidCodesByPrefix(generatedCodes, enteredCode);
+  return getArrayElementsFromIndicies(incomingPublicKeys, validIndicies);
```

Comment:
Why do we need to involve indices? Why not simply:
```ts
export const findPossibleAddressesFromCodePart = (prefix: number[], publicKeys: string[]) => {
  const codes = publicKeys.map((key) => generateCode(key));
  return filterAddressesByPrefix(codes, prefix);
}
```

That should save a lot of lines of code.

Reply:
Client needs an address so he can confirm the request. If you return a code then client would have to go through all notifications to find which address matches up with that returned code. It is ugly to force this on client, no? It is also extra code and more convoluted step, IMHO.